### ISSUE TITLE
Random data generation

### ISSUE THREAD

**Issue by justincampbell** [justincampbell/generative#8]
Using this issue to compile thoughts on this
- Generative should give the user data generation capability out of the box
- Generative should require another gem to accomplish this, as data generation and generative testing are not dependent upon one-another
- There should only be one method to generate data, ie `generate :string` vs `string` or `generate_string`. This is mostly to avoid method name clashes with let/data block names.
- A user should be able to see an example or two in use, and intuitively know how to generate other kinds of data without having to look up documentation

To satisfy these last two, I think I like the idea of having a `generate` method, which takes a symbol and some other options.

I've been thinking about more functionally applicative ways to accomplish data generation. The best example I've found is the [Rantly library](https://github.com/hayeah/rantly#data-generation):

``` rb
Rantly { array(10) { sized(5) { string } } }
```

This is powerful, but I don't find it very intuitive for onboarding people to this process, and I dislike the possibility of name clashes and the uncertainty around that. I'm going to pursue a similar style, but with a single `generate` method:

``` rb
generate(:array, size: 10) { generate :string, size: 5 }
```

This wil be a separate library from Generative, and Generative will require it, but I'm happy to include in our documentation how to include and use other libraries such as Rantly.

/cc @dmcclory @jessitron #3

**Comment by niftyn8** [comment]
Just want to track some thoughts... 

Getting ready soon to implement *something* for shrinking and might take a stab at more composable generators...

I noticed Rantly is doing roughly what I was thinking of for shrinking: https://github.com/hayeah/rantly/blob/master/lib/rantly/shrinks.rb

**Comment by niftyn8** [comment]
More notes... seems like the part that's like rantly is straight forward enough... the tricker part of me is how to hijack the running of an example and rescue from a failure by shrinking and then failing with the shrunk data... Overriding https://github.com/rspec/rspec-core/blob/master/lib/rspec/core/example.rb#L189-L237 seems like a bad idea... not coming up with another path yet...

**Comment by justincampbell** [comment]
@niftyn8 yeah, I also couldn't find a way to access the `let`/`data` from the failure formatter. We could change `data` to store a reference somewhere, but I guess that's not much use if we can't re-run a spec after shrinking.

**Comment by niftyn8** [comment]
@justincampbell some half-baked ideas I haven't really explored yet:

1. Create a subclass of `RSpec::Core::Example`. Something like `GenerativeExample` and write something like an `ExampleGroup.check` method that would replace `it`... Not partial to this idea... Haven't explored it very far but thinking it'd require too much knowledge of RSpec from Generative...

2. Follow Rantly's example and instead ignore example groups and example altogether, using some `check` method inside the block that gets passed to `it`... Haven't fully thought out the consequences but beyond being a definite breaking change it will probably make things like reporting up to RSpec challenging...

**Comment by niftyn8** [comment]
3. Something else involving using a custom method instead of an example group method....
